cirq/devices/graph_device.py

```diff
@@ -0,0 +1,230 @@
+    """An edge of an undirected graph device on which every operation is
+    allowed and has the same duration."""
+
+    def __init__(self, duration: value.Duration):
+        self._duration = duration
+
+
+    def duration_of(self, operation: ops.Operation) -> value.Duration:
+        return self._duration
+
+
+    def validate_operation(self, operation: ops.Operation) -> None:
+        pass
+
+class _UnconstrainedUndirectedGraphDeviceEdge(UndirectedGraphDeviceEdge):
+    """A device edge that allows everything."""
+
+    def duration_of(self, operation: ops.Operation) -> value.Duration:
+        return value.Duration(picos=0)
+
+    def validate_operation(self, operation: ops.Operation) -> None:
+        pass
+
+UnconstrainedUndirectedGraphDeviceEdge = (
+        _UnconstrainedUndirectedGraphDeviceEdge())
+
+
+def is_undirected_device_graph(graph: UndirectedHypergraph) -> bool:
+    if not isinstance(graph, UndirectedHypergraph):
+        return False
+    for _, label in graph.labelled_edges.items():
+        if not (label is None or isinstance(label, UndirectedGraphDeviceEdge)):
+            return False
+    return True
+
+def is_crosstalk_graph(graph: UndirectedHypergraph) -> bool:
+    if not isinstance(graph, UndirectedHypergraph):
+        return False
+    for vertex in graph.vertices:
+        if not isinstance(vertex, frozenset):
+            return False
+    for edge, label in graph.labelled_edges.items():
+        if len(edge) < 2:
+            return False
+        if not ((label is None) or callable(label)):
+            return False
+    return True
+
+
+def raise_crosstalk_error(*ops: ops.Operation):
+    raise ValueError('crosstalk on {}'.format(ops))
+
+
+class UndirectedGraphDevice(devices.Device):
+    """A device whose properties are represented by an edge-labelled graph.
+
+    Each (undirected) edge of the device graph is labelled by an
+    UndirectedGraphDeviceEdge or None. None indicates that any operation is
+    allowed and has zero duration.
+
+    Each (undirected) edge of the constraint graph is labelled either by a
+    function or None. The function takes as arguments operations on the
+    adjacent device edges and raises an error if they are not simultaneously
+    executable. If None, no such operations are allowed.
+
+    Note that
+        * the crosstalk graph is allowed to have vertices (i.e. device edges)
+            that do not exist in the graph device.
+        * duration_of does not check that operation is valid.
+    """
+
+    def __init__(self,
+                 device_graph: UndirectedHypergraph,
+                 crosstalk_graph: Optional[UndirectedHypergraph]=None) -> None:
+        """
+
+        Args:
+            device_graph: An undirected hypergraph whose vertices correspond to
+                qubits and whose edges determine allowable operations and their
+                durations.
+            crosstalk_graph: An undirected hypergraph whose vertices are edges
```

Comment:
What kind of constraints?

Reply:
I would rename it to "disallowed_simultaneous_graph" or something like that, if that's what it is used for.

Reply:
I'm unsure if this is the concept we want, as opposed to a set of pairs of operations that can't be performed simultaneously or a predicate that says yes or no when given a pair of operations. Do you have an example where using the graph is easier than using a predicate lambda or a set?

Reply:
Note that the label of the constraint graph edge allows for specifying a set or predicate of forbidden pairs (or higher-order sets) of operations. By default, `None` indicates that no pair of operations is allowed; that's a good example of where it's easier to use the graph. Another point to make is that with respect to "pairs of operations that can't be performed simultaneously" is that really it's "pairs of operations that can't be performed simultaneously _on adjacent qubits_". The graph is what you the adjacency.

The other advantage of this class is what it exposes to a compiler. Suppose you're a compiler and in the course of doing your business you want to know "where can I do this logical gate?" With the existing predicate-only device class, you'd need to check every possible k-tuple of physical qubits; this class lets you iterate over a much smaller subset of possibilities.